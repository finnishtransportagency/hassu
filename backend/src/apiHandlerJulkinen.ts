import { AppSyncResolverEvent } from "aws-lambda/trigger/appsync-resolver";
import { commonHandleEvent, LambdaResult } from "./api/common";
import { executePublicOperation } from "./api/publicOperations";

export async function handleEvent(event: AppSyncResolverEvent<unknown>): Promise<LambdaResult> {
  return commonHandleEvent(event, executePublicOperation);
}
