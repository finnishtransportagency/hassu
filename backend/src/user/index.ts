export * as userService from "./userService";
export {
  isAorL,
  identifyUser,
  requirePermissionLuonti,
  requirePermissionLuku,
  requirePermissionMuokkaa,
  getVaylaUser,
  requireVaylaUser,
} from "./userService";
