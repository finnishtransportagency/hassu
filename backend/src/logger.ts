import pino, { LogFn } from "pino";
import { getVaylaUser } from "./user";
import { getCorrelationId, reportError } from "./aws/monitoring";

const level = process.env.LOG_LEVEL ? process.env.LOG_LEVEL : "info";
const pretty = process.env.USE_PINO_PRETTY == "true";

function getLogContextOid() {
  return (globalThis as unknown as { oid: string }).oid;
}

export function setLogContextOid(oid: string | undefined): void {
  (globalThis as unknown as { oid: string | undefined }).oid = oid;
}

function getLogger(tag: string) {
  let transport = undefined;
  if (pretty) {
    const isInTest = typeof (global as unknown as Record<string,unknown>).it === "function";
    if (tag == "AUDIT" && isInTest && process.env.TEST_AUDIT_LOG_FILE) {
      transport = { target: "pino/file", options: { destination: process.env.TEST_AUDIT_LOG_FILE } };
    } else {
      transport = {
        target: "pino-pretty",
        options: {
          colorize: true,
          messageFormat: "{tag} {uid} {msg}",
          ignore: "tag,uid",
        },
      };
    }
  }
  // noinspection JSUnusedGlobalSymbols
  return pino({
    level,
    base: undefined,
    mixin: () => {
      return {
        uid: getVaylaUser()?.uid,
        correlationId: getCorrelationId(),
        tag,
        oid: getLogContextOid(),
      };
    },
    formatters: {
      level(label: string) {
        return { level: label };
      },
    },
    timestamp: () => `,"time":"${new Date(Date.now()).toISOString()}"`,
    hooks: {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      logMethod(inputArgs: any[], method: LogFn) {
        for (const inputArg of inputArgs) {
          if (inputArg instanceof Error) {
            reportError(inputArg);
          }
        }
        // Flip parameters to make log("message", {object with fields}) work insted of log({object with fields}, "message")
        if (inputArgs.length >= 2) {
          const arg1 = inputArgs.shift();
          const arg2 = inputArgs.shift();
          return method.apply(this, [arg2, arg1, ...inputArgs]);
        }
        if (inputArgs[0]) {
          return method.apply(this, [inputArgs[0]]);
        }
      },
    },
    transport,
  });
}

export function recordVelhoLatencyDecorator(_target: unknown, propertyKey: string, descriptor: PropertyDescriptor): void {
  // keep a reference to the original function
  const originalValue = descriptor.value;

  // Replace the original function with a wrapper
  descriptor.value = async function (...args: unknown[]) {
    // Call the original function
    const start = Date.now();
    try {
      const result = await originalValue.apply(this, args);
      const end = Date.now();
      logMetricLatency(METRIC_INTEGRATION.VELHO, propertyKey, end - start, true);
      return result;
    } catch (e) {
      log.error(e);
      const end = Date.now();
      logMetricLatency(METRIC_INTEGRATION.VELHO, propertyKey, end - start, false);
      throw e;
    }
  };
}

export enum METRIC_INTEGRATION {
  VELHO = "VELHO",
}

export function logMetricLatency(integration: METRIC_INTEGRATION, operation: string, latency: number, success: boolean): void {
  metricLog.info({
    integration,
    operation,
    latency,
    success,
  });
}

export const log = getLogger("BACKEND");
export const auditLog = getLogger("AUDIT");
const metricLog = getLogger("METRIC");
