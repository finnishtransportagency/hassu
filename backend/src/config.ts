import { log } from "./logger";
import * as AWSXRay from "aws-xray-sdk";

const config = {
  projektiTableName: process.env.TABLE_PROJEKTI,
  projektiArchiveTableName: process.env.TABLE_PROJEKTI_ARCHIVE,
  cognitoURL: process.env.COGNITO_URL,
  velhoAuthURL: process.env.VELHO_AUTH_URL,
  velhoApiURL: process.env.VELHO_API_URL,
  velhoUsername: process.env.VELHO_USERNAME,
  velhoPassword: process.env.VELHO_PASSWORD,

  personSearchApiURL: process.env.PERSON_SEARCH_API_URL,
  personSearchApiURLProd: process.env.PERSON_SEARCH_API_URL_PROD,
  personSearchUsername: process.env.PERSON_SEARCH_API_USERNAME,
  personSearchPassword: process.env.PERSON_SEARCH_API_PASSWORD,
  personSearchAccountTypes: process.env.PERSON_SEARCH_API_ACCOUNT_TYPES?.split(","),
  personSearchUpdaterLambdaArn: process.env.PERSON_SEARCH_UPDATER_LAMBDA_ARN || "",

  searchDomain: process.env.SEARCH_DOMAIN,

  env: process.env.ENVIRONMENT,

  frontendDomainName: process.env.FRONTEND_DOMAIN_NAME,

  frontendPrivateKey: process.env.FRONTEND_PRIVATEKEY,

  uploadBucketName: process.env.UPLOAD_BUCKET_NAME,
  yllapitoBucketName: process.env.YLLAPITO_BUCKET_NAME,
  publicBucketName: process.env.PUBLIC_BUCKET_NAME,
  internalBucketName: process.env.INTERNAL_BUCKET_NAME,
  archiveBucketName: process.env.ARCHIVE_BUCKET_NAME,

  smtpKeyId: process.env.SMTP_KEY_ID,
  smtpSecret: process.env.SMTP_SECRET,

  emailsOn: process.env.EMAILS_ON,
  emailsTo: process.env.EMAILS_TO,
  isProd: (): boolean => process.env.ENVIRONMENT == "prod",
};

process.env.AWS_XRAY_CONTEXT_MISSING = "LOG_ERROR";
AWSXRay.setContextMissingStrategy("IGNORE_ERROR");

// Create your own logger, or instantiate one using a library.
AWSXRay.setLogger({
  error: (message, meta) => {
    log.error(message, { meta });
  },
  warn: (message, meta) => {
    log.warn(message, { meta });
  },
  info: (message, meta) => {
    // Reduce unnecessary logging in local development
    if (process.env.NODE_ENV !== "development") {
      log.info(message, { meta });
    }
  },
  debug: () => {
    // Ignore
  },
});

export { config };
