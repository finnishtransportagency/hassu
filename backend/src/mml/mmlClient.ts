import axios from "axios";
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import { parseStringPromise as parseString } from "xml2js";
import { auditLog, log } from "../logger";
import { chunkArray } from "../database/chunkArray";
import { FeatureCollection } from "geojson";
import lookup from "country-code-lookup";

type Yhteystieto = {
  jakeluosoite?: string | null;
  postinumero?: string | null;
  paikkakunta?: string | null;
  maakoodi?: string | null;
};

export type Omistaja = {
  id?: number;
  kayttooikeusyksikkotunnus?: string;
  etunimet?: string;
  sukunimi?: string;
  nimi?: string;
  henkilotunnus?: string;
  ytunnus?: string;
  kuolinpvm?: string;
  yhteystiedot?: Yhteystieto;
};

export type MmlKiinteisto = {
  kiinteistotunnus?: string;
  kayttooikeusyksikkotunnus?: string;
  omistajat: Omistaja[];
};

export type MmlClient = {
  haeLainhuutotiedot: (kiinteistotunnukset: string[], uid: string, debug?: boolean) => Promise<MmlKiinteisto[]>;
  haeYhteystiedot: (kiinteistotunnukset: string[], uid: string, debug?: boolean) => Promise<MmlKiinteisto[]>;
  haeTiekunnat: (kiinteistotunnukset: string[], uid: string, debug?: boolean) => Promise<MmlKiinteisto[]>;
  haeYhteisalueet: (kiinteistotunnukset: string[], uid: string, debug?: boolean) => Promise<MmlKiinteisto[]>;
};

export type MmlOptions = {
  ogcEndpoint: string;
  endpoint: string;
  apiKey: string;
  ogcApiKey: string;
  ogcApiExamples: string;
};

const TIMEOUT = 120000;
const MAX = 100;
const CONCURRENT_MAX = 10;

export function getMmlClient(options: MmlOptions): MmlClient {
  return {
    haeLainhuutotiedot: async (kiinteistotunnukset: string[], uid: string, debug = false) => {
      const all: MmlKiinteisto[] = [];
      const chunks = chunkArray(kiinteistotunnukset, MAX);
      for (const chunk of chunks) {
        const responses: Promise<MmlKiinteisto[]>[] = [];
        for (const kyselytunnukset of chunkArray(chunk, CONCURRENT_MAX)) {
          responses.push(
            new Promise((resolve: (yhteystiedot: MmlKiinteisto[]) => void) => {
              const yhteystiedot: MmlKiinteisto[] = [];
              const url = options.endpoint + "/lainhuutotiedot/xml?kiinteistotunnus=" + kyselytunnukset.join(",");
              log.info("haeLainhuutotiedot url: " + url);
              auditLog.info("Lainhuutotietojen haku", { kiinteistotunnukset: kyselytunnukset });
              const vastaustunnukset = [...kyselytunnukset];
              axios.get(url, { headers: { "x-api-key": options.apiKey, enduserid: uid }, timeout: TIMEOUT }).then((response) => {
                if (debug) {
                  console.log("rawdata: %s", response.data.replaceAll("\n", ""));
                }
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                parseString(response.data).then((responseJson: any) => {
                  if (responseJson["kylh:Lainhuutotiedot"]["kylh:Rekisteriyksikko"]) {
                    for (const yksikko of responseJson["kylh:Lainhuutotiedot"]["kylh:Rekisteriyksikko"]) {
                      const omistajat: Omistaja[] = [];
                      const kiinteistotunnus = yksikko["trpt:rekisteriyksikonPerustiedot"][0]["y:kiinteistotunnus"][0];
                      if (yksikko["trlh:lainhuudot"][0]["trlh:Lainhuutoasia"]) {
                        for (const asia of yksikko["trlh:lainhuudot"][0]["trlh:Lainhuutoasia"]) {
                          // Ratkaistu tai Loppuunsaatettu
                          if ((asia["y:asianTila"][0] === "02" || asia["y:asianTila"][0] === "03") && asia["trlh:osuudetAsianKohteesta"]) {
                            for (const kohde of asia["trlh:osuudetAsianKohteesta"][0]["trlh:OsuusAsianKohteesta"]) {
                              for (const hlo of kohde["y:osuudenHenkilot"][0]["y:Henkilo"]) {
                                const tiedot = hlo["y:henkilonTiedot"][0];
                                omistajat.push({
                                  henkilotunnus: tiedot["y:henkilotunnus"] ? tiedot["y:henkilotunnus"][0] : undefined,
                                  etunimet: tiedot["y:etunimet"] ? tiedot["y:etunimet"][0] : undefined,
                                  sukunimi: tiedot["y:sukunimi"] ? tiedot["y:sukunimi"][0] : undefined,
                                  ytunnus: tiedot["y:ytunnus"] ? tiedot["y:ytunnus"][0] : undefined,
                                  nimi: tiedot["y:nimi"] ? tiedot["y:nimi"][0] : undefined,
                                  kuolinpvm: tiedot["y:kuolinpvm"] ? tiedot["y:kuolinpvm"][0] : undefined,
                                });
                              }
                            }
                          }
                        }
                      }
                      const idx = vastaustunnukset.indexOf(kiinteistotunnus);
                      if (idx !== -1) {
                        vastaustunnukset.splice(idx, 1);
                      }
                      yhteystiedot.push({ kiinteistotunnus, omistajat });
                    }
                  }
                  for (const kiinteistotunnus of vastaustunnukset) {
                    yhteystiedot.push({ kiinteistotunnus, omistajat: [] });
                  }
                  resolve(yhteystiedot);
                });
              });
            })
          );
        }
        all.push(...(await Promise.all(responses)).flat());
      }
      return all;
    },
    haeYhteystiedot: async (kiinteistotunnukset: string[], uid: string, debug = false) => {
      const all: MmlKiinteisto[] = [];
      const chunks = chunkArray(kiinteistotunnukset, MAX);
      for (const chunk of chunks) {
        const responses: Promise<MmlKiinteisto[]>[] = [];
        for (const kyselytunnukset of chunkArray(chunk, CONCURRENT_MAX)) {
          responses.push(
            new Promise((resolve: (yhteystiedot: MmlKiinteisto[]) => void) => {
              const yhteystiedot: MmlKiinteisto[] = [];
              const url = options.endpoint + "/yhteystiedot/xml?kiinteistotunnus=" + kyselytunnukset.join(",");
              log.info("haeYhteystiedot url: " + url);
              auditLog.info("Yhteystietojen haku", { kiinteistotunnukset: kyselytunnukset });
              const vastaustunnukset = [...kyselytunnukset];
              axios.get(url, { headers: { "x-api-key": options.apiKey, enduserid: uid }, timeout: TIMEOUT }).then((response) => {
                if (debug) {
                  console.log("rawdata: %s", response.data.replaceAll("\n", ""));
                }
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                parseString(response.data).then((responseJson: any) => {
                  if (responseJson["kyyh:Yhteystiedot"]["tryh:Rekisteriyksikko"]) {
                    for (const yksikko of responseJson["kyyh:Yhteystiedot"]["tryh:Rekisteriyksikko"]) {
                      const omistajat: Omistaja[] = [];
                      const kiinteistotunnus = yksikko["trpt:rekisteriyksikonPerustiedot"][0]["y:kiinteistotunnus"][0];
                      if (yksikko["tryh:kohteenHenkilot"][0]["y:Henkilo"]) {
                        for (const hlo of yksikko["tryh:kohteenHenkilot"][0]["y:Henkilo"]) {
                          const tiedot = hlo["y:henkilonTiedot"][0];
                          const osoite = hlo["muti:Osoite"] ? hlo["muti:Osoite"][0] : undefined;
                          omistajat.push({
                            etunimet: tiedot["y:etunimet"] ? tiedot["y:etunimet"][0] : undefined,
                            sukunimi: tiedot["y:sukunimi"] ? tiedot["y:sukunimi"][0] : undefined,
                            nimi: tiedot["y:nimi"] ? tiedot["y:nimi"][0] : undefined,
                            yhteystiedot: {
                              jakeluosoite: osoite && osoite["muti:jakeluosoite"] ? osoite["muti:jakeluosoite"][0] : undefined,
                              paikkakunta: osoite && osoite["muti:paikkakunta"] ? osoite["muti:paikkakunta"][0] : undefined,
                              postinumero: osoite && osoite["muti:postinumero"] ? osoite["muti:postinumero"][0] : undefined,
                              maakoodi: osoite && osoite["muti:maakoodi"] ? lookup.byIso(osoite["muti:maakoodi"][0])?.iso2 : undefined,
                            },
                          });
                        }
                      }
                      const idx = vastaustunnukset.indexOf(kiinteistotunnus);
                      if (idx !== -1) {
                        vastaustunnukset.splice(idx, 1);
                      }
                      yhteystiedot.push({ kiinteistotunnus, omistajat });
                    }
                    for (const kiinteistotunnus of vastaustunnukset) {
                      yhteystiedot.push({ kiinteistotunnus, omistajat: [] });
                    }
                    resolve(yhteystiedot);
                  }
                });
              });
            })
          );
        }
        all.push(...(await Promise.all(responses)).flat());
      }
      return all;
    },
    haeTiekunnat: async (kiinteistotunnukset, uid, debug = false) => {
      const yhteystiedot: MmlKiinteisto[] = [];
      const chunks = chunkArray(kiinteistotunnukset, MAX);
      const tiekuntaMap = new Map<number, Omistaja>();
      for (const kyselytunnukset of chunks) {
        let url =
          options.ogcEndpoint +
          "/collections/KayttooikeusyksikonPerustiedot/items?kiinteistotunnus=" +
          kyselytunnukset.join(",") +
          "&tiekunnallisuus=1";
        auditLog.info("Tiekuntien haku", { kiinteistotunnukset: kyselytunnukset });
        let response = await axios.get(url, {
          headers: { "x-api-key": options.ogcApiKey, enduserid: uid },
          timeout: TIMEOUT,
        });
        if (debug) {
          console.log("rawdata: %s", JSON.stringify(response.data));
        }
        let geojson = response.data as FeatureCollection | undefined;
        const ids: number[] = [];
        for (const feat of geojson?.features || []) {
          if (feat.properties?.tiekunta) {
            feat.properties.tiekunta.forEach((t: { nimi: string; id: number }) => {
              const omistaja = { id: t.id, nimi: t.nimi, kayttooikeusyksikkotunnus: feat.properties?.kayttooikeusyksikkotunnus };
              if (!tiekuntaMap.has(t.id)) {
                tiekuntaMap.set(t.id, omistaja);
                ids.push(t.id);
              }
            });
          }
        }
        url = options.ogcEndpoint + "/collections/TiekunnanYhteystiedot/items?id=" + ids.join(",");
        if (options.ogcApiExamples === "true") {
          url = url.replace("/features/", "/examples/");
        }
        auditLog.info("Tiekuntien yhteystietojen haku", { ids, url });
        response = await axios.get(url, {
          headers: { "x-api-key": options.ogcApiKey, enduserid: uid },
          timeout: TIMEOUT,
        });
        if (debug) {
          console.log("rawdata: %s", JSON.stringify(response.data));
        }
        geojson = response.data as FeatureCollection | undefined;
        for (const feat of geojson?.features ?? []) {
          if (feat?.properties?.yhteyshenkilo[0]) {
            const omistaja = tiekuntaMap.get(feat.id as number);
            if (omistaja && !omistaja.yhteystiedot) {
              // tiekunnan nimi sulkuihin
              omistaja.nimi = feat.properties.yhteyshenkilo[0].nimi + (omistaja.nimi ? " (" + omistaja.nimi + ")" : "");
              omistaja.yhteystiedot = {
                jakeluosoite: feat.properties.yhteyshenkilo[0].osoite[0]?.osoite,
                postinumero: feat.properties.yhteyshenkilo[0].osoite[0]?.postinumero,
                paikkakunta: feat.properties.yhteyshenkilo[0].osoite[0]?.postitoimipaikka,
                maakoodi: feat.properties.yhteyshenkilo[0].osoite[0]?.valtio
                  ? lookup.byIso(feat.properties.yhteyshenkilo[0].osoite[0].valtio)?.iso2
                  : undefined,
              };
            }
          }
        }
      }
      for (const yhteyshenkilo of tiekuntaMap.values()) {
        yhteystiedot.push({
          kayttooikeusyksikkotunnus: yhteyshenkilo.kayttooikeusyksikkotunnus,
          omistajat: [{ nimi: yhteyshenkilo.nimi, yhteystiedot: yhteyshenkilo.yhteystiedot }],
        });
      }
      return yhteystiedot;
    },
    haeYhteisalueet: async (kiinteistotunnukset, uid, debug = false) => {
      const yhteystiedot: MmlKiinteisto[] = [];
      const chunks = chunkArray(kiinteistotunnukset, MAX);
      const alueMap = new Map<string, MmlKiinteisto>();
      for (const kyselytunnukset of chunks) {
        const url =
          options.ogcEndpoint + "/collections/RekisteriyksikonYhteisalueosuudet/items?kiinteistotunnus=" + kyselytunnukset.join(",");
        auditLog.info("Yhteisalueiden haku", { kiinteistotunnukset: kyselytunnukset });
        const response = await axios.get(url, {
          headers: { "x-api-key": options.ogcApiKey, enduserid: uid },
          timeout: TIMEOUT,
        });
        if (debug) {
          console.log("rawdata: %s", JSON.stringify(response.data));
        }
        const geojson = response.data as FeatureCollection;
        for (const feat of geojson.features) {
          for (const alue of feat.properties?.yhteisalueosuus ?? []) {
            if (!alueMap.has(alue.yhteisalueyksikko.kiinteistotunnus)) {
              alueMap.set(alue.yhteisalueyksikko.kiinteistotunnus, {
                kiinteistotunnus: alue.yhteisalueyksikko.kiinteistotunnus,
                omistajat: [{ nimi: alue.yhteisalueyksikko.nimi }],
              });
            }
          }
        }
      }
      yhteystiedot.push(...alueMap.values());
      return yhteystiedot;
    },
  };
}
