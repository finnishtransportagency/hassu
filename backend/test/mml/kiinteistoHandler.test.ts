import { SQSRecord } from "aws-lambda";
import {
  DBOmistaja,
  OmistajaHakuEvent,
  haeKiinteistonOmistajat,
  handleEvent,
  poistaKiinteistonOmistaja,
  setClient,
  tallennaKiinteistonOmistajat,
} from "../../src/mml/kiinteistoHandler";
import { MmlClient } from "../../src/mml/mmlClient";
import { BatchGetCommand, BatchWriteCommand, DynamoDBDocumentClient, GetCommand, PutCommand, UpdateCommand } from "@aws-sdk/lib-dynamodb";
import { assert, expect } from "chai";
import { mockClient } from "aws-sdk-client-mock";
import { setLogContextOid } from "../../src/logger";
import { identifyMockUser } from "../../src/user/userService";
import { config } from "../../src/config";

const mockMmlClient: MmlClient = {
  haeLainhuutotiedot: (_kiinteistotunnukset: string[]) => {
    return new Promise((resolve) => {
      resolve([
        {
          kiinteistotunnus: "1",
          omistajat: [
            {
              henkilotunnus: "111111-111A",
              etunimet: "Matti",
              sukunimi: "Testaaja",
              yhteystiedot: { jakeluosoite: "Katuosoite 1 A 100", postinumero: "00100", paikkakunta: "Helsinki" },
            },
            {
              henkilotunnus: "111111-113B",
              etunimet: "Teppo",
              sukunimi: "Testaaja",
              yhteystiedot: { jakeluosoite: "Katuosoite 2 B 32", postinumero: "01600", paikkakunta: "Vantaa" },
            },
          ],
        },
        {
          kiinteistotunnus: "2",
          omistajat: [
            {
              henkilotunnus: "222222-222A",
              etunimet: "Lotta",
              sukunimi: "Testaaja",
              yhteystiedot: { jakeluosoite: "Katuosoite 3 A 123", postinumero: "00180", paikkakunta: "Helsinki" },
            },
          ],
        },
        {
          kiinteistotunnus: "3",
          omistajat: [
            {
              etunimet: "Jatta",
              sukunimi: "Tuntematon",
              yhteystiedot: { jakeluosoite: "Katuosoite 123", postinumero: "00180", paikkakunta: "Helsinki" },
            },
          ],
        },
        {
          kiinteistotunnus: "4",
          omistajat: [],
        },
      ]);
    });
  },
};

describe("kiinteistoHandler", () => {
  after(() => {
    setClient(undefined);
  });

  before(() => {
    setClient(mockMmlClient);
  });

  beforeEach(() => {
    setLogContextOid("1");
    identifyMockUser({ etunimi: "", sukunimi: "", uid: "testuid", __typename: "NykyinenKayttaja" });
  });

  it("tallenna kiinteistön omistajat", async () => {
    const event: OmistajaHakuEvent = { oid: "1", uid: "test", kiinteistotunnukset: ["1", "2", "3", "4"] };
    const record: SQSRecord = { body: JSON.stringify(event) } as unknown as SQSRecord;
    const dbMock = mockClient(DynamoDBDocumentClient);
    await handleEvent({ Records: [record] });
    dbMock.on(BatchWriteCommand).resolves({});
    expect(dbMock.commandCalls(BatchWriteCommand).length).to.be.equal(1);
    const writeCommand = dbMock.commandCalls(BatchWriteCommand)[0];
    assert(writeCommand.args[0].input.RequestItems);
    expect(writeCommand.args[0].input.RequestItems[config.omistajaTableName].length).to.be.equal(5);
    expect(dbMock.commandCalls(UpdateCommand).length).to.be.equal(1);
    const updateCommand = dbMock.commandCalls(UpdateCommand)[0];
    assert(updateCommand.args[0].input.ExpressionAttributeValues);
    expect(updateCommand.args[0].input.ExpressionAttributeValues[":omistajat"].length).to.be.equal(3);
    expect(updateCommand.args[0].input.ExpressionAttributeValues[":muutOmistajat"].length).to.be.equal(2);
  });
  it("poista kiinteistön omistaja", async () => {
    const dbMock = mockClient(DynamoDBDocumentClient);
    dbMock.on(GetCommand).resolves({ Item: { id: "1", omistajat: ["1"], muutOmistajat: ["2"] } });
    await poistaKiinteistonOmistaja({ oid: "1", omistaja: "1" });
    expect(dbMock.commandCalls(UpdateCommand).length).to.be.equal(1);
    const updateCommand = dbMock.commandCalls(UpdateCommand)[0];
    assert(updateCommand.args[0].input.ExpressionAttributeValues);
    expect(updateCommand.args[0].input.ExpressionAttributeValues[":omistajat"].length).to.be.equal(0);
    expect(updateCommand.args[0].input.ExpressionAttributeValues[":muutOmistajat"].length).to.be.equal(1);
    expect(updateCommand.args[0].input.ExpressionAttributeValues[":muutOmistajat"][0]).to.be.equal("2");
    dbMock.reset();
    dbMock.on(GetCommand).resolves({ Item: { id: "1", omistajat: ["11"], muutOmistajat: ["22"] } });
    await poistaKiinteistonOmistaja({ oid: "1", omistaja: "22" });
    expect(dbMock.commandCalls(UpdateCommand).length).to.be.equal(1);
    const updateCommand2 = dbMock.commandCalls(UpdateCommand)[0];
    assert(updateCommand2.args[0].input.ExpressionAttributeValues);
    expect(updateCommand2.args[0].input.ExpressionAttributeValues[":omistajat"].length).to.be.equal(1);
    expect(updateCommand2.args[0].input.ExpressionAttributeValues[":muutOmistajat"].length).to.be.equal(0);
    expect(updateCommand2.args[0].input.ExpressionAttributeValues[":omistajat"][0]).to.be.equal("11");
  });
  it("päivitä kiinteistön omistajat", async () => {
    const dbMock = mockClient(DynamoDBDocumentClient);
    dbMock.on(GetCommand, { TableName: config.projektiTableName }).resolves({ Item: { id: "1" } });
    dbMock.on(GetCommand, { TableName: config.omistajaTableName, Key: { id: "11" } }).resolves({ Item: { id: "11", etunimet: "Teppo" } });
    await tallennaKiinteistonOmistajat({
      oid: "1",
      omistajat: [
        {
          id: "11",
          kiinteistotunnus: "1",
          etunimet: "Teppo2",
          sukunimi: "Tepon sukunimi",
          yhteystiedot: { jakeluosoite: "Osoite 2", postinumero: "00100", paikkakunta: "Helsinki" },
        },
        {
          kiinteistotunnus: "2",
          etunimet: "Matti",
          sukunimi: "Ruohonen",
          yhteystiedot: { jakeluosoite: "Osoite 1", postinumero: "01000", paikkakunta: "Vantaa" },
        },
      ],
    });
    expect(dbMock.commandCalls(PutCommand).length).to.be.equal(2);
    const putCommand = dbMock.commandCalls(PutCommand)[0];
    const o1 = putCommand.args[0].input.Item as DBOmistaja;
    expect(o1.etunimet).to.be.equal("Teppo2");
    expect(o1.sukunimi).to.be.equal("Tepon sukunimi");
    expect(o1.yhteystiedot?.jakeluosoite).to.be.equal("Osoite 2");
    expect(o1.yhteystiedot?.postinumero).to.be.equal("00100");
    expect(o1.yhteystiedot?.paikkakunta).to.be.equal("Helsinki");
    const putCommand2 = dbMock.commandCalls(PutCommand)[1];
    const o2 = putCommand2.args[0].input.Item as DBOmistaja;
    expect(o2.etunimet).to.be.equal("Matti");
    expect(o2.sukunimi).to.be.equal("Ruohonen");
    expect(o2.yhteystiedot?.jakeluosoite).to.be.equal("Osoite 1");
    expect(o2.yhteystiedot?.postinumero).to.be.equal("01000");
    expect(o2.yhteystiedot?.paikkakunta).to.be.equal("Vantaa");
  });
  it("hae kiinteistön omistajat", async () => {
    const dbMock = mockClient(DynamoDBDocumentClient);
    dbMock
      .on(GetCommand, { TableName: config.projektiTableName })
      .resolves({ Item: { id: "1", omistajat: ["1", "2", "3"], muutOmistajat: ["4"] } });
    dbMock.on(BatchGetCommand).resolves({
      Responses: {
        [config.omistajaTableName]: [
          { id: "1", etunimet: "Matti", yhteystiedot: { jakeluosoite: "Osoite 1", postinumero: "00100", paikkakunta: "Helsinki" } },
          { id: "2", etunimet: "Teppo" },
        ],
      },
    });
    const omistajat = await haeKiinteistonOmistajat({ oid: "1", muutOmistajat: false, sivu: 1, sivuKoko: 2 });
    expect(dbMock.commandCalls(BatchGetCommand).length).to.be.equal(1);
    let batchCommand = dbMock.commandCalls(BatchGetCommand)[0];
    assert(batchCommand.args[0].input.RequestItems);
    let keys = batchCommand.args[0].input.RequestItems[config.omistajaTableName].Keys;
    assert(keys);
    expect(keys.length).to.be.equal(2);
    expect(keys[0].id).to.be.equal("1");
    expect(keys[1].id).to.be.equal("2");
    expect(omistajat.hakutulosMaara).to.be.equal(3);
    expect(omistajat.sivu).to.be.equal(1);
    expect(omistajat.sivunKoko).to.be.equal(2);
    expect(omistajat.omistajat[0]?.id).to.be.equal("1");
    expect(omistajat.omistajat[0]?.etunimet).to.be.equal("Matti");
    expect(omistajat.omistajat[0]?.yhteystiedot?.jakeluosoite).to.be.equal("Osoite 1");
    expect(omistajat.omistajat[0]?.yhteystiedot?.postinumero).to.be.equal("00100");
    expect(omistajat.omistajat[0]?.yhteystiedot?.paikkakunta).to.be.equal("Helsinki");
    expect(omistajat.omistajat[1]?.id).to.be.equal("2");
    expect(omistajat.omistajat[1]?.etunimet).to.be.equal("Teppo");
    await haeKiinteistonOmistajat({ oid: "1", muutOmistajat: false, sivu: 2, sivuKoko: 2 });
    expect(dbMock.commandCalls(BatchGetCommand).length).to.be.equal(2);
    batchCommand = dbMock.commandCalls(BatchGetCommand)[1];
    assert(batchCommand.args[0].input.RequestItems);
    keys = batchCommand.args[0].input.RequestItems[config.omistajaTableName].Keys;
    assert(keys);
    expect(keys.length).to.be.equal(1);
    expect(keys[0].id).to.be.equal("3");
  });
});
