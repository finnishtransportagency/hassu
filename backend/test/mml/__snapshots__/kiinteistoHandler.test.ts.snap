// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kiinteistoHandler tallenna kiinteist√∂n omistajat 1`] = `
Array [
  Object {
    "etunimet": "Matti",
    "expires": 0,
    "henkilotunnus": "111111-111A",
    "id": "0",
    "kiinteistotunnus": "1",
    "lisatty": "",
    "nimi": undefined,
    "oid": "1.2.3",
    "sukunimi": "Testaaja",
    "yhteystiedot": Object {
      "jakeluosoite": "Katuosoite 1 A 100",
      "maakoodi": "FI",
      "paikkakunta": "Helsinki",
      "postinumero": "00100",
    },
    "ytunnus": undefined,
  },
  Object {
    "etunimet": "Teppo",
    "expires": 0,
    "henkilotunnus": "111111-113B",
    "id": "1",
    "kiinteistotunnus": "1",
    "lisatty": "",
    "nimi": undefined,
    "oid": "1.2.3",
    "sukunimi": "Testaaja",
    "yhteystiedot": Object {
      "jakeluosoite": "Katuosoite 2 B 32",
      "maakoodi": undefined,
      "paikkakunta": "Vantaa",
      "postinumero": "01600",
    },
    "ytunnus": undefined,
  },
  Object {
    "etunimet": "Lotta",
    "expires": 0,
    "henkilotunnus": "222222-222A",
    "id": "2",
    "kiinteistotunnus": "2",
    "lisatty": "",
    "nimi": undefined,
    "oid": "1.2.3",
    "sukunimi": "Testaaja",
    "yhteystiedot": Object {
      "jakeluosoite": "Katuosoite 3 A 123",
      "maakoodi": undefined,
      "paikkakunta": "Helsinki",
      "postinumero": "00180",
    },
    "ytunnus": undefined,
  },
  Object {
    "etunimet": "Jatta",
    "expires": 0,
    "henkilotunnus": undefined,
    "id": "3",
    "kiinteistotunnus": "3",
    "lisatty": "",
    "nimi": undefined,
    "oid": "1.2.3",
    "sukunimi": "Tuntematon",
    "yhteystiedot": Object {
      "jakeluosoite": "Katuosoite 123",
      "maakoodi": undefined,
      "paikkakunta": "Helsinki",
      "postinumero": "00180",
    },
    "ytunnus": undefined,
  },
  Object {
    "etunimet": undefined,
    "expires": 0,
    "henkilotunnus": undefined,
    "id": "4",
    "kiinteistotunnus": "4",
    "lisatty": "",
    "nimi": "Yritys Oy Ab",
    "oid": "1.2.3",
    "sukunimi": undefined,
    "yhteystiedot": Object {
      "jakeluosoite": "Yritysosoite 1",
      "maakoodi": undefined,
      "paikkakunta": "Helsinki",
      "postinumero": "00001",
    },
    "ytunnus": "123-456",
  },
  Object {
    "expires": 0,
    "id": "5",
    "kiinteistotunnus": "5",
    "lisatty": "",
    "oid": "1.2.3",
  },
]
`;
